File: windows/SeedSeeker/LevelMap.cs
// SPDX-License-Identifier: GPL-3.0-or-later
using System.Text.Json;
using System.Text.Json.Nodes;

namespace SeedSeeker;

/// <summary>The v3 scene is authoritative; contents metadata never becomes extra drawing.</summary>
public sealed class LevelMapDocument
{
    public string Format { get; init; } = "";
    public int SchemaVersion { get; init; }
    public string Profile { get; init; } = "";
    public string AssetRevision { get; init; } = "";
    public string Seed { get; init; } = "";
    public int Depth { get; init; }
    public int Branch { get; init; }
    public string Kind { get; init; } = "";
    public string? SelectedTrinket { get; init; }
    public int Width { get; init; }
    public int Height { get; init; }
    public int[][] SecretRooms { get; init; } = [];
    public int[] SecretDoors { get; init; } = [];
    public int[] SecretTraps { get; init; } = [];
    public MapBranch[] Branches { get; init; } = [];
    public MapAsset[] Assets { get; init; } = [];
    public MapScene Scene { get; init; } = new();
    public bool HasSecrets => SecretRooms.Length + SecretDoors.Length + SecretTraps.Length > 0;

    public static LevelMapDocument Parse(byte[] json)
    {
        var map = JsonSerializer.Deserialize<LevelMapDocument>(json, new JsonSerializerOptions { PropertyNameCaseInsensitive = true })
            ?? throw new InvalidDataException("The engine returned an empty map.");
        if (map.Format != "seed-seeker-level-map" || map.SchemaVersion != 3)
            throw new InvalidDataException("This map format requires a newer version of Seed Seeker.");
        if (map.Width <= 0 || map.Height <= 0 || map.Width > 256 || map.Height > 256 || map.Scene.TileSize != 16
            || map.Scene.Layers.Length == 0 || map.Scene.ConcealedLayers.Length == 0
            || map.Scene.Layers.Concat(map.Scene.ConcealedLayers).Any(layer => layer.Cells.Length != map.Width * map.Height
                || layer.Cells.Any(index => index is int i && (i < 0 || i >= map.Scene.Sprites.Length)))
            || map.Scene.Sprites.Any(sprite => sprite.Frames.Length == 0 || sprite.FrameDurationMs <= 0))
            throw new InvalidDataException("The engine returned an invalid map scene.");
        return map;
    }

    public static string Request(string seed, int depth, int branch, QuerySettings query, string? trinket)
    {
        var request = new JsonObject
        {
            ["seed"] = seed, ["depth"] = depth, ["branch"] = branch,
            ["challenges"] = new JsonArray([.. Challenges.All.Where(c => (query.Challenges & c.Mask) != 0).Select(c => JsonValue.Create(c.Name))]),
            ["trinket"] = trinket,
        };
        if (query.Requirements.Count > 0) request["query"] = JsonNode.Parse(ResultsExport.EncodeQueryDocument(query));
        return request.ToJsonString();
    }
}

public sealed record MapBranch(int Depth, int Branch, string Kind, int Entrance);
public sealed record MapAsset(string Id, int Width, int Height, string Sha256);
public sealed class MapScene
{
    public int TileSize { get; init; }
    public MapSprite[] Sprites { get; init; } = [];
    public MapLayer[] Layers { get; init; } = [];
    public MapLayer[] ConcealedLayers { get; init; } = [];
    public MapEmitter[] Emitters { get; init; } = [];
    public MapEmitter[] ConcealedEmitters { get; init; } = [];
}
public sealed record MapLayer(string Name, string? Blend, int?[] Cells);
public sealed record MapSprite(double FrameDurationMs, MapDraw[][] Frames);
public sealed record MapGlow(int[] Color, double PeriodMs);
public sealed class MapDraw
{
    public string Kind { get; init; } = "";
    public string Asset { get; init; } = "";
    public double[] Source { get; init; } = [];
    public double[] Destination { get; init; } = [];
    public int[] Rgba { get; init; } = [];
    public int Opacity { get; init; } = 255;
    public int[]? Tint { get; init; }
    public MapGlow? Glow { get; init; }
}
public sealed record MapCurve(double[][] Points, bool Sqrt)
{
    public double Value(double progress)
    {
        var p = Math.Clamp(progress, 0, 1) * 1000;
        var right = Array.FindIndex(Points, point => point[0] >= p);
        double value;
        if (right <= 0) value = Points[right == 0 ? 0 : ^1][1];
        else
        {
            var a = Points[right - 1]; var b = Points[right];
            value = a[1] + (b[1] - a[1]) * (p - a[0]) / (b[0] - a[0]);
        }
        return Sqrt ? Math.Sqrt(Math.Max(0, value / 1000)) : value / 1000;
    }
}
public sealed class MapEmitter
{
    public double? StartMs { get; init; }
    public bool WallMask { get; init; }
    public bool ClipToChasm { get; init; }
    public int Cell { get; init; }
    public double LoopMs { get; init; }
    public string? Blend { get; init; }
    public MapDraw Image { get; init; } = new();
    public double[] Velocity { get; init; } = [0, 0];
    public double[] Acceleration { get; init; } = [0, 0];
    public double AngularSpeed { get; init; }
    public MapCurve Alpha { get; init; } = new([[0, 1000]], false);
    public MapCurve Scale { get; init; } = new([[0, 1000]], false);
    public MapCurve? ScaleX { get; init; }
    public MapCurve? ScaleY { get; init; }
    public MapParticle[] Particles { get; init; } = [];

    public MapParticleState? State(MapParticle particle, double elapsed)
    {
        var clock = Math.Max(0, elapsed) - (StartMs ?? 0);
        if (StartMs.HasValue && clock < particle.BirthMs) return null;
        var age = ((clock - particle.BirthMs) % LoopMs + LoopMs) % LoopMs;
        if (age >= particle.LifespanMs) return null;
        var seconds = age / 1000; var progress = age / particle.LifespanMs;
        return new(particle.Position[0] / 1000 + Velocity[0] * seconds + Acceleration[0] * seconds * seconds / 2,
            particle.Position[1] / 1000 + Velocity[1] * seconds + Acceleration[1] * seconds * seconds / 2,
            particle.Scale / 1000 * Scale.Value(progress), ScaleX?.Value(progress) ?? 1, ScaleY?.Value(progress) ?? 1,
            Alpha.Value(progress), (particle.Angle + AngularSpeed * seconds) * Math.PI / 180);
    }
}
public sealed record MapParticle(double BirthMs, double LifespanMs, double[] Position, double Scale, double Angle);
public sealed record MapParticleState(double X, double Y, double Scale, double ScaleX, double ScaleY, double Alpha, double Angle);

/// <summary>Choice letters and conflicts follow the jointly obtainable engine match.</summary>
public static class ScoutChoices
{
    public static string Letter(int group) => ((char)('A' + group % 26)).ToString();
    public static IReadOnlyDictionary<int, ulong> Matched(IReadOnlyList<ScoutItem> items, IReadOnlySet<int> matches) =>
        items.Select((item, index) => (item, index)).Where(x => matches.Contains(x.index) && x.item.AccessibilityTag == 1)
            .GroupBy(x => x.item.AccessibilityGroup).ToDictionary(g => g.Key, g => g.Last().item.AccessibilityValue);
    public static bool Dimmed(ScoutItem item, bool matched, IReadOnlyDictionary<int, ulong> choices) =>
        !matched && item.AccessibilityTag == 1 && choices.TryGetValue(item.AccessibilityGroup, out var option) && option != item.AccessibilityValue;
}

File: macos/SeedSeeker/Sources/SeedSeekerKit/LevelMap.swift
import CSeedFinder
import Foundation

/// A scene clock uses uptime independently of its reduced-motion sample. When
/// motion resumes, scheduled hazards begin at their captured first turn.
public struct LevelMapClock: Sendable {
    private var origin: Double
    private var wasReduced = false
    public init(now: Double) { origin = now }
    public mutating func elapsed(at now: Double, reducedMotion: Bool) -> Double {
        if wasReduced && !reducedMotion { origin = now }
        wasReduced = reducedMotion
        return reducedMotion ? 0 : max(0, now - origin)
    }
}

/// The resolved scout profile, independent of query edits and view state.
public struct LevelMapRequest: Hashable, Sendable {
    public let seed: String
    public let depth: Int
    public let branch: Int
    public let challenges: Int
    public let selectedTrinket: String?

    public init(seed: String, depth: Int, branch: Int = 0, challenges: Int, selectedTrinket: String?) {
        self.seed = seed; self.depth = depth; self.branch = branch
        self.challenges = challenges; self.selectedTrinket = selectedTrinket
    }

    /// Trinket changes refresh a scene without navigating away from its map.
    public func hasSameLocation(as other: Self) -> Bool {
        seed == other.seed && depth == other.depth && branch == other.branch && challenges == other.challenges
    }

    public func encoded() throws -> Data {
        guard (0...SearchLimits.challengeMask).contains(challenges) else { throw SeedFinderEngineError.invalidArgument }
        let names = EngineInfo.shared.challengeNames.filter { $0.key & challenges != 0 }.map(\.value).sorted()
        return try JSONSerialization.data(withJSONObject: [
            "seed": seed, "depth": depth, "branch": branch,
            "challenges": names,
            "trinket": selectedTrinket ?? "none",
        ], options: [.sortedKeys])
    }
}

/// Version 3 is a complete scene: contents, secrets and effects are drawn from
/// the selected layer stack without reconstructing any game rules in Swift.
public struct LevelMapDocument: Decodable, Sendable {
    public let format: String
    public let schemaVersion: Int
    public let assetRevision: String
    public let seed: String
    public let depth: Int
    public let branch: Int
    public let kind: String
    public let selectedTrinket: String?
    public let width: Int
    public let height: Int
    public let secretRooms: [[Int]]
    public let secretDoors: [Int]
    public let secretTraps: [Int]
    public let branches: [Branch]
    public let assets: [Asset]
    public let scene: Scene
    public let itemTooltips: [ItemTooltip]?

    public struct ItemTooltip: Decodable, Sendable {
        public let cell: Int
        public let label: String
        public let hidden: Bool
        public let items: [TooltipItem]
        public let bounds: [Int]?
    }
    public struct TooltipItem: Decodable, Sendable {
        public let name: String
        public let description: String
        public let image: Int
        public let quantity: Int
        public let deterministic: Bool
        public let icon: [Int]?
    }
    public func itemAt(x: Double, y: Double, secrets: Bool) -> ItemTooltip? {
        guard x >= 0, y >= 0, x < Double(pixelWidth), y < Double(pixelHeight) else { return nil }
        return itemTooltips?.first { tip in
            let bounds = tip.bounds ?? [0, 0, scene.tileSize, scene.tileSize]
            let left = Double((tip.cell % width) * scene.tileSize + bounds[0])
            let top = Double((tip.cell / width) * scene.tileSize + bounds[1])
            return (secrets || !tip.hidden) && x >= left && y >= top && x < left + Double(bounds[2]) && y < top + Double(bounds[3])
        }
    }

    public var secretCount: Int { secretRooms.count + secretDoors.count + secretTraps.count }
    public var pixelWidth: Int { width * scene.tileSize }
    public var pixelHeight: Int { height * scene.tileSize }

    public struct Branch: Decodable, Sendable, Identifiable {
        public let depth: Int
        public let branch: Int
        public let kind: String
        public var id: Int { branch }
        public var label: String { kind == "imp_vault" ? "Imp Vault" : "Blacksmith Mine" }
    }
    public struct Asset: Decodable, Sendable {
        public let id: String
        public let width: Int
        public let height: Int
        public let sha256: String
    }
    public struct Scene: Decodable, Sendable {
        public let tileSize: Int
        public let sprites: [Sprite]
        public let layers: [Layer]
        public let concealedLayers: [Layer]
        public let emitters: [Emitter]
        public let concealedEmitters: [Emitter]
    }
    public struct Layer: Decodable, Sendable {
        public let name: String
        public let blend: String?
        public let cells: [Int?]
    }
    public struct Sprite: Decodable, Sendable {
        public let frameDurationMs: Double
        public let frames: [[Draw]]
        public func frameIndex(at elapsed: Double) -> Int {
            Int(max(0, elapsed) / frameDurationMs) % frames.count
        }
    }
    public struct Draw: Decodable, Hashable, Sendable {
        public let kind: String
        public let destination: [Double]
        public let asset: String?
        public let source: [Double]?
        public let opacity: Double?
        public let tint: [Double]?
        public let glow: Glow?
        public let rgba: [Double]?
    }
    public struct Glow: Decodable, Hashable, Sendable {
        public let color: [Double]
        public let periodMs: Double
        public func strength(at elapsed: Double) -> Double {
            let phase = max(0, elapsed).truncatingRemainder(dividingBy: periodMs * 2) / periodMs
            return min(phase, 2 - phase) * 0.6
        }
    }
    public struct Curve: Decodable, Sendable {
        public let points: [[Double]]
        public let sqrt: Bool
        public func value(at progress: Double) -> Double {
            let p = min(1, max(0, progress)) * 1000
            let value: Double
            if let right = points.firstIndex(where: { $0[0] >= p }) {
                if right == 0 { value = points[0][1] }
                else {
                    let a = points[right - 1], b = points[right]
                    value = a[1] + (b[1] - a[1]) * (p - a[0]) / (b[0] - a[0])
                }
            } else { value = points.last![1] }
            return sqrt ? Foundation.sqrt(max(0, value / 1000)) : value / 1000
        }
    }
    public struct Particle: Decodable, Sendable {
        public let birthMs: Double
        public let lifespanMs: Double
        public let position: [Double]
        public let scale: Double
        public let angle: Double
    }
    public struct ParticleState: Sendable {
        public let x: Double, y: Double, scale: Double, scaleX: Double, scaleY: Double, alpha: Double, angle: Double
    }
    public struct Emitter: Decodable, Sendable {
        public let startMs: Double?
        public let wallMask: Bool?
        public let clipToChasm: Bool?
        public let cell: Int
        public let loopMs: Double
        public let blend: String?
        public let image: Draw
        public let velocity: [Double]
        public let acceleration: [Double]
        public let angularSpeed: Double
        public let alpha: Curve
        public let scale: Curve
        public let scaleX: Curve?
        public let scaleY: Curve?
        public let particles: [Particle]

        public func state(of particle: Particle, at elapsed: Double) -> ParticleState? {
            let clock = max(0, elapsed) - (startMs ?? 0)
            if startMs != nil && clock < particle.birthMs { return nil }
            let age = ((clock - particle.birthMs).truncatingRemainder(dividingBy: loopMs) + loopMs)
                .truncatingRemainder(dividingBy: loopMs)
            guard age < particle.lifespanMs else { return nil }
            let seconds = age / 1000, progress = age / particle.lifespanMs
            return ParticleState(
                x: particle.position[0] / 1000 + velocity[0] * seconds + acceleration[0] * seconds * seconds / 2,
                y: particle.position[1] / 1000 + velocity[1] * seconds + acceleration[1] * seconds * seconds / 2,
                scale: particle.scale / 1000 * scale.value(at: progress),
                scaleX: scaleX?.value(at: progress) ?? 1,
                scaleY: scaleY?.value(at: progress) ?? 1, alpha: alpha.value(at: progress),
                angle: (particle.angle + angularSpeed * seconds) * .pi / 180)
        }
    }

    public static func decode(_ data: Data) throws -> Self {
        let map = try JSONDecoder().decode(Self.self, from: data)
        guard map.format == "seed-seeker-level-map", map.schemaVersion == 3,
              (1...256).contains(map.width), (1...256).contains(map.height), map.scene.tileSize == 16
        else { throw SeedFinderEngineError.invalidResponse }
        let count = map.width * map.height
        guard
              (map.scene.layers + map.scene.concealedLayers).allSatisfy({ layer in
                  layer.cells.count == count && layer.cells.allSatisfy { $0 == nil || map.scene.sprites.indices.contains($0!) }
              }), map.scene.sprites.allSatisfy({ $0.frameDurationMs > 0 && !$0.frames.isEmpty })
        else { throw SeedFinderEngineError.invalidResponse }
        let assetIDs = Set(map.assets.map(\.id))
        func validDraw(_ draw: Draw) -> Bool {
            guard draw.destination.count == 4 else { return false }
            if draw.kind == "fill" { return draw.rgba?.count == 4 }
            return draw.kind == "blit" && assetIDs.contains(draw.asset ?? "") && draw.source?.count == 4
                && (draw.tint == nil || draw.tint?.count == 3)
                && (draw.glow == nil || (draw.glow!.color.count == 3 && draw.glow!.periodMs > 0))
        }
        func validCurve(_ curve: Curve) -> Bool {
            !curve.points.isEmpty && curve.points.allSatisfy { $0.count == 2 }
                && zip(curve.points, curve.points.dropFirst()).allSatisfy { $0[0] < $1[0] }
        }
        guard map.scene.sprites.allSatisfy({ $0.frames.allSatisfy { $0.allSatisfy(validDraw) } }),
              (map.scene.emitters + map.scene.concealedEmitters).allSatisfy({ emitter in
                  (0..<count).contains(emitter.cell) && emitter.loopMs > 0 && validDraw(emitter.image)
                      && emitter.velocity.count == 2 && emitter.acceleration.count == 2
                      && validCurve(emitter.alpha) && validCurve(emitter.scale)
                      && (emitter.scaleX == nil || validCurve(emitter.scaleX!))
                      && (emitter.scaleY == nil || validCurve(emitter.scaleY!))
                      && emitter.particles.allSatisfy { $0.position.count == 2 && $0.lifespanMs > 0 }
              }) else { throw SeedFinderEngineError.invalidResponse }
        return map
    }
}

public struct LevelMapBundle: Sendable {
    public let document: LevelMapDocument
    public let assets: [String: Data]
}

/// Native generation and asset reads execute on this actor, away from the UI.
/// One engine is statically linked for the process lifetime; its revision is
/// immutable. PNGs are shared by content digest, maps use an eight-entry LRU.
public actor LevelMapClient {
    public static let shared = LevelMapClient()
    private var maps: [LevelMapRequest: LevelMapBundle] = [:]
    private var order: [LevelMapRequest] = []
    private var assets: [String: Data] = [:]

    public func load(_ request: LevelMapRequest) throws -> LevelMapBundle {
        try Task.checkCancellation()
        if let cached = maps[request] {
            order.removeAll { $0 == request }; order.append(request)
            return cached
        }
        let input = try request.encoded()
        let packet = try enginePacket { out, length in
            input.withUnsafeBytes { bytes in
                seedfinder_level_map(bytes.bindMemory(to: UInt8.self).baseAddress, bytes.count, out, length)
            }
        }
        let map = try LevelMapDocument.decode(packet)
        var images: [String: Data] = [:]
        for asset in map.assets {
            if assets[asset.sha256] == nil {
                let id = Data(asset.id.utf8)
                assets[asset.sha256] = try enginePacket { out, length in
                    id.withUnsafeBytes { bytes in
                        seedfinder_level_map_asset(bytes.bindMemory(to: UInt8.self).baseAddress, bytes.count, out, length)
                    }
                }
            }
            images[asset.id] = assets[asset.sha256]
        }
        let bundle = LevelMapBundle(document: map, assets: images)
        maps[request] = bundle; order.append(request)
        while order.count > 8 { maps.removeValue(forKey: order.removeFirst()) }
        return bundle
    }
}

/// Match witnesses select an option for a whole group, including all its items.
/// Unmatched items in the selected option must stay readable.
public struct ScoutChoiceStatus: Sendable {
    private let selected: [Int: Int]
    public init(items: [ScoutItem], matched: Set<Int>) {
        var choices: [Int: Int] = [:]
        for (index, item) in items.enumerated() where matched.contains(index) {
            if case let .choice(group, option) = item.accessibility { choices[group] = option }
        }
        selected = choices
    }
    public func isDimmed(_ accessibility: ScoutAccessibility, matched: Bool) -> Bool {
        guard !matched, case let .choice(group, option) = accessibility, let chosen = selected[group] else { return false }
        return chosen != option
    }
    public static func letter(_ group: Int) -> String {
        String(UnicodeScalar(65 + max(0, group) % 26)!)
    }
}
